nextcloud: add option to set datadir and extensions

beardhatcode opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
I want to manage my Nextcloud extensions trough nix and move extensions out of the datadir. 

This moves the data for apps managed by nix to a `nix-apps` folder

Feedback is welcome, my config is:
```nix
let  mkNCApp = pkgs.fetchNextcloudApp;
in {
        services.nextcloud = {
          enable = true;
          hostName = nextcloudSub;
          package = nc;
          webfinger = true;
          https = true;
          datadir = "/mnt/datafile";
          extraApps = {
            maps =
              mkNCApp rec {
                name = "maps";
                version = "0.1.8";
                url = "https://github.com/nextcloud/maps/releases/download/v${version}/maps-${version}.tar.gz";
                sha256 = "d7ffb425fa0f11d2c3b7cad857eca258c94d635d77663fe524a6e6bff5416c41";
              };
            calendar =
              mkNCApp rec {
                name = "calendar";
                version = "2.1.2";
                url = "https://github.com/nextcloud/calendar/releases/download/v${version}/calendar.tar.gz";
                sha256 = "rt3pPdDFnd67K5K/zlScUkEN/+fOTgbNTR6hbA2sQuc=";
              };
            tasks =
              mkNCApp rec {
                name = "tasks";
                version = "0.13.6";
                url = "https://github.com/nextcloud/tasks/releases/download/v${version}/tasks.tar.gz";
                sha256 = "mLvnvp4DFa8kdXLabvLG9n7hOHBlymmlaYjuCnc2fxE=";
              };
            contacts =
              mkNCApp rec {
                name = "contacts";
                version = "3.4.2";
                url = "https://github.com/nextcloud/contacts/releases/download/v${version}/contacts.tar.gz";
                sha256 = "y3JCbAu204uxmqugkbZm+9F8J2g7WtyCh+lmlu53+Uw=";
              };
            notes =
              mkNCApp rec {
                name = "notes";
                version = "4.0.2";
                url = "https://github.com/nextcloud/notes/releases/download/v${version}/notes.tar.gz";
                sha256 = "LxJ3J0n8Bvan5sWNvBZiD//PPuNNnmZ8rAn/Pnj4WKs=";
              };
            previewgenerator =
              mkNCApp rec {
                name = "previewgenerator";
                version = "3.1.0";
                url = "https://github.com/rullzer/previewgenerator/releases/download/v${version}/previewgenerator.tar.gz";
                sha256 = "TdDdzPPuqRqaLqZB0aq4BzX7grX6xuaAlj15/9zd63Q=";
              };

          };
          caching = {
            apcu = true;
          };
          phpOptions = {
            catch_workers_output = "yes";
            display_errors = "stderr";
            error_reporting = "E_ALL & ~E_DEPRECATED & ~E_STRICT";
            expose_php = "Off";
            "opcache.enable_cli" = "1";
            "opcache.fast_shutdown" = "1";
            "opcache.interned_strings_buffer" = "8";
            "opcache.max_accelerated_files" = "10000";
            "opcache.memory_consumption" = "128";
            "opcache.revalidate_freq" = "1";
            short_open_tag = "Off";
          };
          poolSettings = {
            pm = "dynamic";
            "pm.max_children" = "120";
            "pm.max_requests" = "500";
            "pm.max_spare_servers" = "18";
            "pm.min_spare_servers" = "6";
            "pm.start_servers" = "12";
          };
          config = {
            dbtype = "pgsql";
            dbuser = "nextcloud";
            dbhost = "/run/postgresql"; # nextcloud will add /.s.PGSQL.5432 by itself
            dbname = "nextcloud";
            adminpass = "wasabi4001isacoolpasswordbutyoushouldpickabetterone";
            adminuser = "root";
            trustedProxies = [ hostIP ];
          };
        };
}
  ```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

chvp commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -76,6 +77,16 @@ in {
+    extraApps = mkOption {
+      type = types.attrsOf types.package;
+      default = { };
+      description = "Extra apps to install";
```
An example would probably be useful here.

chvp commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -644,6 +667,10 @@ in {
             priority = 201;
             extraConfig = "root ${cfg.home};";
           };
+            "~ ^/nix-apps" = {
+              priority = 201;
+              extraConfig = "root ${cfg.home};";
+            };
```
This indentation looks off.

chvp commented:
It might be nice to provide a builder like the `mkNCApp` from your config in nixpkgs.

beardhatcode commented:
@chvp where should such a builder go?

chvp commented:
Probably in a file next to nextcloud's default.nix? I'm not entirely sure to be honest.

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -76,6 +77,16 @@ in {
     };
+    datadir = mkOption {
+      type = types.str;
+      default = null;
```
An example would be nice.

beardhatcode commented:
@Ma27, do you have opinions on this?

Ma27 commented:
Idea seems OK on a first look. Have you checked how Nextcloud behaves if you try to add or update apps that were installed with Nix?

beardhatcode commented:
Yes, I updated the calender app this morning and installed phonetrack last month.
But it is true that I need to check how this behaves if there are migrations (if there even is such a thing).

beardhatcode commented:
I'm calling `${occ}/bin/nextcloud-occ upgrade` after the nix based Nextcloud apps are put in place, that should carry out any required migrations.

https://github.com/NixOS/nixpkgs/blob/ad0e917fef78838548430f9a35bd3e2dd8fa03d1/nixos/modules/services/web-apps/nextcloud.nix#L591

Ma27 commented:
Sounds OK then. I hope you understand that I'd still like to test all of that on my own as maintainer of this module. I'll try to do it today or tomorrow. In case I forget about this, don't hesitate to ping me :)

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -76,6 +77,17 @@ in {
     };
+    datadir = mkOption {
+      type = types.str;
+      default = null;
```
If you set it to `null` by default, you have to use `types.nullOr types.str` as `type`. Otherwise you'll get an error like this:

```
error: A definition for option `services.nextcloud.datadir' is not of type `string'. Definition values:
       - In `/home/ma27/Projects/nixpkgs/nixos/modules/services/web-apps/nextcloud.nix': null
(use '--show-trace' to show detailed location information)

```

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -76,6 +77,17 @@ in {
+    extraApps = mkOption {
+      type = types.attrsOf types.package;
+      default = { };
+      description = "Extra apps to install. Should point to packages that have appinfo/info.xml in their root.";
```
To me it's not entirely clear how the attribute-name has to look like. I assume that `maps = ...` where the repository is also named `maps`, correct? It would be good if that would be documented here. if the description becomes too long, you may want to use multi-line strings btw:

```
description = ''
  very
  long
  description
'';
```

beardhatcode replied:
done

Ma27 commented on pkgs/build-support/fetchnextcloudapp/default.nix:
```diff
@@ -0,0 +1,43 @@
+{ stdenv, pkgs, ... }:
```
Please don't use `pkgs` here. Instead you can specify `gnutar`, `findutils` etc. explicitly. This also has the benefit that one could override these dependencies like this:

```
fetchNextcloudApp.override { findutils = myfindutis; }
```

Ma27 commented on pkgs/build-support/fetchnextcloudapp/default.nix:
```diff
@@ -0,0 +1,43 @@
+
+  src = pkgs.fetchurl {
+    url = url;
+    sha256 = sha256;
```
I still think that these `foo = foo` expressions are kinda weird. Instead you'd generally use `inherit foo`.

nlewo replied:
Yep, we prefer to use `inherit`.

Ma27 commented on pkgs/build-support/fetchnextcloudapp/default.nix:
```diff
@@ -0,0 +1,43 @@
+  installPhase = ''
+    tgr=$(dirname $(dirname $(find -path '*/appinfo/info.xml' | head -n 1)))
+
+    if [ \! -d $tgr ];
```
I may be misunderstanding something here, but why is the backslash needed here?

beardhatcode replied:
no, it has been removed now

Ma27 commented on pkgs/build-support/fetchnextcloudapp/default.nix:
```diff
@@ -0,0 +1,43 @@
+    findutils
+  ];
+
+  buildInputs = with pkgs; [ ];
```
If no buildInputs are needed, this can be omitted.

Ma27 reviewed (commented):
Changes look mostly good, just a few notes below.

Regarding the implementation itself, I'm afraid I misunderstood how to use this. I just wrote a VM expression like this:

``` nix
{
  nextcloud = { lib, pkgs, config, ... }: {
    networking.firewall.allowedTCPPorts = [ 80 ];
    services.nextcloud = {
      enable = true;
      hostName = "nextcloud";
      config.adminpass = "foobar";
      datadir = "/var/lib/nextcloud";
      extraApps.maps = pkgs.fetchNextcloudApp rec {
        name = "maps";
        version = "0.1.8";
        url = "https://github.com/nextcloud/maps/releases/download/v${version}/maps-${version}.tar.gz";
        sha256 = "d7ffb425fa0f11d2c3b7cad857eca258c94d635d77663fe524a6e6bff5416c41";
      };
    };
  };
}
```

(I built it via `NIX_PATH=nixpkgs=$(pwd) nixos-build-vms vm.nix` and then ran `QEMU_NET_OPTS="hostfwd=tcp::8080-:80" ./result/bin/nixos-run-vms` to access it in my browser on `localhost:8080`).

From what I've seen, this should make the Maps app available when I login, correct? Unfortunately the top-menu only gives me Files/Folders/Activity, but no Maps. To be sure, I tried to enable it in the administrator panel. When I hit `Enable` for the (now-available) Maps app, I get the error `The "unique" column option is not supported.`. In the logs, the following message appears:

```
nextcloud # [   91.528339] Nextcloud[907]: {"reqId":"OYqtGMnPVHwRZRwJKLT9","level":3,"time":"2021-05-02T17:42:04+00:00","remoteAddr":"10.0.2.2","user":"root","app":"settings","method":"POST","url":"/settings/apps/enable","message":"{\"Exception\":\"Doctrine\\\\DBAL\\\\Schema\\\\Exception\\\\UnknownColumnOption\",\"Message\":\"The \\\"unique\\\" column option is not supported.\",\"Code\":0,\"Trace\":[{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/3rdparty/doctrine/dbal/src/Schema/Column.php\",\"line\":84,\"function\":\"new\",\"class\":\"Doctrine\\\\DBAL\\\\Schema\\\\Exception\\\\UnknownColumnOption\",\"type\":\"::\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/3rdparty/doctrine/dbal/src/Schema/Column.php\",\"line\":68,\"function\":\"setOptions\",\"class\":\"Doctrine\\\\DBAL\\\\Schema\\\\Column\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/3rdparty/doctrine/dbal/src/Schema/Table.php\",\"line\":355,\"function\":\"__construct\",\"class\":\"Doctrine\\\\DBAL\\\\Schema\\\\Column\",\"type\":\"->\"},{\"file\":\"/nix/store/firwnx70whrqfycdn92iiks5pm428qac-nc-app-maps/lib/Migration/Version000009Date20190625000800.php\",\"line\":43,\"function\":\"addColumn\",\"class\":\"Doctrine\\\\DBAL\\\\Schema\\\\Table\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/lib/private/DB/MigrationService.php\",\"line\":455,\"function\":\"changeSchema\",\"class\":\"OCA\\\\Maps\\\\Migration\\\\Version000009Date20190625000800\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/lib/private/DB/MigrationService.php\",\"line\":418,\"function\":\"migrateSchemaOnly\",\"class\":\"OC\\\\DB\\\\MigrationService\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/lib/private/Installer.php\",\"line\":163,\"function\":\"migrate\",\"class\":\"OC\\\\DB\\\\MigrationService\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/apps/settings/lib/Controller/AppSettingsController.php\",\"line\":448,\"function\":\"installApp\",\"class\":\"OC\\\\Installer\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/lib/private/AppFramework/Http/Dispatcher.php\",\"line\":218,\"function\":\"enableApps\",\"class\":\"OCA\\\\Settings\\\\Controller\\\\AppSettingsController\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/lib/private/AppFramework/Http/Dispatcher.php\",\"line\":127,\"function\":\"executeController\",\"class\":\"OC\\\\AppFramework\\\\Http\\\\Dispatcher\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/lib/private/AppFramework/App.php\",\"line\":157,\"function\":\"dispatch\",\"class\":\"OC\\\\AppFramework\\\\Http\\\\Dispatcher\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/lib/private/Route/Router.php\",\"line\":302,\"function\":\"main\",\"class\":\"OC\\\\AppFramework\\\\App\",\"type\":\"::\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/lib/base.php\",\"line\":993,\"function\":\"match\",\"class\":\"OC\\\\Route\\\\Router\",\"type\":\"->\"},{\"file\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/index.php\",\"line\":37,\"function\":\"handleRequest\",\"class\":\"OC\",\"type\":\"::\"}],\"File\":\"/nix/store/kk42k7xk0mc4d42z2j0hy2c3s035kgf7-nextcloud-21.0.1/3rdparty/doctrine/dbal/src/Schema/Exception/UnknownColumnOption.php\",\"Line\":18,\"CustomMessage\":\"--\"}","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:88.0) Gecko/20100101 Firefox/88.0","version":"21.0.1.1"}
```

Now I'm wondering, did I do anything wrong here?

Last but not least, we should document this new feature in `nixos/modules/services/web-apps/nextcloud.xml`. I know a lot of people don't like docbook, so as soon as everything else is done, I can take care of this as well :)

beardhatcode commented:
I've been trying to find out the cause for this issue @Ma27, It appears to be a bug in the maps app ( https://github.com/nextcloud/maps/issues/541 ) if you are working with SqlLite and in my tests I've been using Postgres.

This vm.nix works for me: (only chane is the bumped version of maps)
```nix
{
  nextcloud = { lib, pkgs, config, ... }: {
    networking.firewall.allowedTCPPorts = [ 80 ];
    services.nextcloud = {
      enable = true;
      hostName = "nextcloud";
      config.adminpass = "foobar";
      datadir = "/var/lib/nextcloud";
      extraApps.maps = pkgs.fetchNextcloudApp rec {
        name = "maps";
        version = "0.1.9-8-nightly";
        url = "https://github.com/nextcloud/maps/releases/download/v${version}/maps-${version}.tar.gz";
        sha256 = "sha256-3uLg1rvSDcgQgUvqhNsu1v08+72jIzvyo20ckyBWJ6w=";
      };
    };
  };
}
```

Kranzes commented:
What is the state of this PR? I am unable to test it at the moment, I'd like to know from your experience.

Ma27 commented:
I'm not really sure what the state is either, but it seems readier than I had in mind, so I can start a second attempt to test soonish.

Kranzes commented:
Sounds good!

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -547,21 +560,32 @@ in {
+              lib.mapAttrsToList
+                (name: target:
+                  "ln -sf   ${target} ${cfg.home}/nix-apps/${name}")
+                cfg.extraApps) +
```
I think it may be nicer, easier to read and perhaps less error-prone to use `buildEnv` to create a store-path with all apps inside. This can be replaced atomically then in `nextcloud-setup` by replacing the symlink on `${cfg.home}/nix-apps` only.

beardhatcode replied:
I did it with [`linkFarm`](https://github.com/NixOS/nixpkgs/blob/4be93515614857491efe5ec00eba8105b893cce3/pkgs/build-support/trivial-builders.nix#L360) instead. `buildEnv` joined the contents of the apps in one folder

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -547,21 +560,32 @@ in {
 
             ln -sf ${cfg.package}/apps ${cfg.home}/
+            rm -rf ${cfg.home}/nix-apps
+            mkdir -p ${cfg.home}/nix-apps
```
Can you explain why you're using this points to `home` rather than `datadir`?

beardhatcode replied:
Because store-apps is there. To me (personally) it did not make sense to put the apps in the datadir because they are not really data (to me). My Nextcloud is in an ephemeral container so with datadir I can wipe everything except the data on every update. Where do you think the apps should go @Ma27 ?

Ma27 replied:
Fair enough :+1:

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -79,6 +80,17 @@ in {
+    datadir = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = "Data storage path of nextcloud. Will be services.nextcloud.home by default. This folder will be populated with a config and data folder which contain the state of the instance (excl the database).";
```
This should really be a multi-line description IMHO.
Also I'd prefer to have options referenced like `<xref linkend="opt-services.nextcloud.home" />`

Ma27 commented on pkgs/build-support/fetchnextcloudapp/default.nix:
```diff
@@ -0,0 +1,39 @@
+    if [ -d "$approot" ];
+    then
+      mv "$approot/" $out
+      chmod -R g-w $out
```
This shouldn't be needed. `/nix/store` is read-only anyways.

beardhatcode replied:
I added this because nix was complaining about "suspicious permissions" (or something like that)

beardhatcode replied:
I've been experimenting with this a bit, and it seems that I even need:
```suggestion
      chmod -R a-w $out
```

For the tasks app (since its most recent version)

Ma27 replied:
It seems that this is done by Nix to avoid having other processes writing into a store-path that's about to be created and with `cp` the permissions would be different. As suggested above, the `chmod` should be sufficient to fix this entirely.

beardhatcode replied:
Done in 98f95f9

Ma27 reviewed (commented):
* Tested it in a VM with `maps-0.1.9` which works perfectly fine :+1: 
* It would be nice if you could document this in `nixos/modules/services/web-apps/nextcloud.xml`. It may be especially confusing that you have to enable these declarative modules manually in the admin-interface.

Ma27 commented on pkgs/build-support/fetchnextcloudapp/default.nix:
```diff
@@ -0,0 +1,39 @@
+
+  unpackPhase = ''
+    tar -xzpf $src
+    find -type f -executable -exec chmod o-x,g-x-w,a-x-w '{}' \;
```
I think the easiest solution is a `chmod -R go-w` here to leave it out below in `installPhase` entirely :)

beardhatcode replied:
Done in 98f95f9

Ma27 reviewed (commented):
LGTM, mostly.

First of all, thanks a lot for your hard work on this!

To get this  merged, the following things should be done:
* please write some documentation about how this works (and that apps have to be enabled by an administrator before being able to use them) into `nixos/modules/services/web-apps/nextcloud.xml`. If you need help with docbook, just ping me! :) 
* Please fix the remaining review-notes.
* Please squash the commits into logical junks and make sure that the commit-messages adhere our contribution-guidelines. Also, please remove my GH handle from the messages as these cause unnecessary notifications for me whenever someone pushes them (or rebases a branch).

Ma27 commented:
@beardhatcode do you intend to continue working on this? :)

beardhatcode commented:
Yes, I do as I'm still using it, but I don't have much time at the moment.

Ma27 commented:
@beardhatcode what's the status here? :)

beardhatcode commented:
I don't have much time to work on this. Currently, my plan is to finish this PR on 25 September :sweat_smile:

beardhatcode committed:
nixos/nextcloud: add services.nextcloud.datadir setting

This option can be used to set an alternative storage location for
files and app metadata.

beardhatcode committed:
nixos/nextcloud: add settings to manage nextcloud apps

Note the appstoreEnable which will prevent nextcloud form updating
nix-managed apps. This is needed because nextcloud will store an other
version of the app in /var/lib/nextcloud/store-apps and it will
no longer be manageable.

beardhatcode commented:
@Ma27 , I finished this up today and squashed the commits.

I also added an option to disable the appstore. It is enabled by default when the `extraApps` feature is used and prevents Nextcloud from updating these apps.

nlewo commented on pkgs/build-support/fetchnextcloudapp/default.nix:
```diff
@@ -0,0 +1,38 @@
+  '';
+
+  installPhase = ''
+    approot="$(dirname $(dirname $(find -path '*/appinfo/info.xml' | head -n 1)))"
```
Is there an app which doesn't have the `appinfo` folder in the archive root?
We are [deploying several apps](https://framagit.org/markas/infrastructure/-/blob/7e65a9050750d5e46c9fc93f1ac9af65821bea49/modules/nextcloud.nix#L10) with Nix and didn't see this kind of issues.

We could then use the `fetchzip` fetcher instead.

beardhatcode replied:
During testing I noticed that some apps were placed in a folder and had a bunch of non-related stuff in the zip. This removed that. (I can't remember which app is was tho)

nlewo commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -548,31 +558,37 @@ in {
             ${occ}/bin/nextcloud-occ config:system:delete trusted_domains
+
+            ${optionalString (cfg.extraAppsEnable && cfg.extraApps != { }) ''
+                # Try to enable apps (don't fail when one of them cannot be enabled , eg. due to incompatible version)
```
Hm, i'm not sure this is a good default behavior: i have NixOS tests for my infrastructure and i would prefer to get a systemd failure when an app is not compatible.

We could actually ensure app are compatible with the nextcloud version at build time by reading the `appinfo/info.xml` application file! (it's out of the scope of this PR:))

nlewo reviewed (commented):
Thanks @beardhatcode ! That's pretty cool!

We actually have this feature on a [shared infra](https://framagit.org/markas/infrastructure/-/blob/7e65a9050750d5e46c9fc93f1ac9af65821bea49/modules/nextcloud.nix#L9) but we didn't find/take time to upstream it. So, i'm glad you are upstreaming this!

It would be really nice to enrich the Nextcloud NixOS test. I think we could add a simple app (such as the calendar one) and do a basic test to ensure it is working as expected (see [here](https://framagit.org/markas/infrastructure/-/blob/7e65a9050750d5e46c9fc93f1ac9af65821bea49/tests/nextcloud.nix#L172) for instance). Do not hesitate to ping me (lewo on Matrix/IRC) if you need help on that topic.

beardhatcode committed:
nixos/nextcloud: fail systemd-service if apps don't activate

beardhatcode commented:
@nlewo I was thinking of an even simpler app that would just expose some unprotected endpoint and could be compactly written in the test file.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.xml:
```diff
@@ -237,6 +237,12 @@
   </para>
+
+  <para>
+   Alternatively, extra apps can also be declared in with the <xref linkend="opt-services.nextcloud.extraApps" /> setting.
```
"be declared with the"

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -79,6 +80,59 @@ in {
     };
+    datadir = mkOption {
+      type = types.nullOr types.str;
+      default = null;
```
It would be slightly more elegant to set this to `cfg.home` by default, then you can get rid of the `if/else` at the top.

However, please do so by declaring `defaultText` in the option and `datadir = mkOptionDefault cfg.home` in the config-block. That way one doesn't have to rebuild the manual if this option gets changed.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -460,10 +514,16 @@ in {
+                then '''appstoreenabled' => ${lib.boolToString cfg.appstoreEnable},''
+                else (if (cfg.extraApps != { })
+                      then '''appstoreenabled' => false,''
+                      else "")}
```
Why don't we set this value at all here? Wouldn't it make sense to use upstream's default (i.e. `true`/`false`) here?

beardhatcode replied:
Because this may break existing configurations. If users have set `appstoreenabled' => false,` in their own `config.php`, our `override.config.php` will enable their appstore. I opted to only override the property if it is explicitly set by the user (by setting `appstoreEnable` or using `extraApps`)

nlewo commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -79,6 +80,59 @@ in {
+      example = "/mnt/nextcloud-file";
+    };
+    extraApps = mkOption {
+      type = types.attrsOf types.package;
```
In the future, it would be nice to provide a set of tested Nextcloud apps. The user could then choose to enable a set of prepackaged Nextcloud apps or add its own. I'm not sure the current type would be convenient to choose an app amongst a predefined list.

A type such as `extraApps = types.listOf types.package` would then allow us to create the following type in the fututre `types.listOf (types.Or types.package (types.enum ['calendar' 'contact']))`.

Currently, the attribute set key is only used to define the app directory name. What about using the `id` value found in the `appinfo/info.xml` file instead?

Ma27 committed:
nixos/nextcloud: fix wording in manual

Ma27 committed:
nixos/nextcloud: fixup #119638

A few minor changes to get #119638 - nextcloud: add option to set
datadir and extensions - ready:

* `cfg.datadir` now gets `cfg.home` as default to make the type
  non-nullable.
* Enhanced the `basic` test to check the behavior with a custom datadir
  that's not `/var/lib/nextcloud`.
* Fix hashes for apps in option example.
* Simplify if/else for `appstoreenable` in override config.
* Simplify a few `mapAttrsToList`-expressions in
  `nextcloud-setup.service`.

Ma27 committed:
nixos/nextcloud: build manual without warnings
